Code_Exercises/In_Order_Queue/solution1.cpp

```diff
@@ -0,0 +1,101 @@
+#include <sycl.hpp>
+
+#define MAD_4(x, y)                                                                                \
+  x = y * x + y;                                                                                   \
+  y = x * y + x;                                                                                   \
+  x = y * x + y;                                                                                   \
+  y = x * y + x;
+#define MAD_16(x, y)                                                                               \
+  MAD_4(x, y);                                                                                     \
+  MAD_4(x, y);                                                                                     \
+  MAD_4(x, y);                                                                                     \
+  MAD_4(x, y);
+#define MAD_64(x, y)                                                                               \
+  MAD_16(x, y);                                                                                    \
+  MAD_16(x, y);                                                                                    \
+  MAD_16(x, y);                                                                                    \
+  MAD_16(x, y);
+
+template <class T> static T busy_wait(size_t N, T i) {
+  T x = 1.3f;
+  T y = i;
+  for (size_t j = 0; j < N; j++) {
+    MAD_64(x, y);
+  }
+  return y;
+}
+
+auto bench(sycl::queue Q, int N_kernel) {
+  auto *out = sycl::malloc_device<double>(N_kernel, Q);
+
+  const auto s = std::chrono::high_resolution_clock::now();
+  for (int i = 0; i < N_kernel; i++) {
+    Q.single_task([=]() { out[i] = busy_wait<double>(1E5, i); });
```

Comment:
Do we know that all the hardware available at IWOCL will have double precision support?

Reply:
I can use `float` no problem.